cdc/sink/dispatcher/event_router.go

```diff
@@ -56,27 +57,92 @@ func (r *partitionDispatchRule) fromString(rule string) {
 // EventRouter is a router, it determines which topic and which partition
 // an event should be dispatched to.
 type EventRouter struct {
-	rules []struct {
+	defaultTopic string
+	rules        []struct {
 		partitionDispatcher partition.Dispatcher
 		topicDispatcher     topic.Dispatcher
 		filter.Filter
 	}
 }
 
+// NewEventRouter creates a new EventRouter
+func NewEventRouter(cfg *config.ReplicaConfig, defaultTopic string) (*EventRouter, error) {
+	// If an event does not match any dispatching rules in the config file,
+	// it will be dispatched by the default partition dispatcher and
+	// static topic dispatcher because it matches *.* rule.
+	ruleConfigs := append(cfg.Sink.DispatchRules, &config.DispatchRule{
+		Matcher:       []string{"*.*"},
+		PartitionRule: "default",
+		TopicRule:     "",
+	})
+	rules := make([]struct {
+		partitionDispatcher partition.Dispatcher
+		topicDispatcher     topic.Dispatcher
+		filter.Filter
+	}, 0, len(ruleConfigs))
+
+	for _, ruleConfig := range ruleConfigs {
+		f, err := filter.Parse(ruleConfig.Matcher)
+		if err != nil {
+			return nil, cerror.WrapError(cerror.ErrFilterRuleInvalid, err)
+		}
+		if !cfg.CaseSensitive {
+			f = filter.CaseInsensitive(f)
+		}
+
+		d := getPartitionDispatcher(ruleConfig, cfg.EnableOldValue)
+		t, err := getTopicDispatcher(ruleConfig, defaultTopic)
+		if err != nil {
+			return nil, err
+		}
+		rules = append(rules, struct {
+			partitionDispatcher partition.Dispatcher
+			topicDispatcher     topic.Dispatcher
+			filter.Filter
+		}{partitionDispatcher: d, topicDispatcher: t, Filter: f})
+	}
+	return &EventRouter{
+		defaultTopic: defaultTopic,
+		rules:        rules,
+	}, nil
+}
+
 // DispatchRowChangedEvent returns the target topic and partition.
-func (s *EventRouter) DispatchRowChangedEvent(row *model.RowChangedEvent) (string, int32) {
-	topicDispatcher, partitionDispatcher := s.matchDispatcher(row)
+func (s *EventRouter) DispatchRowChangedEvent(row *model.RowChangedEvent, partitionNum int32) (string, int32) {
+	topicDispatcher, partitionDispatcher := s.matchDispatcher(row.Table.Schema, row.Table.Table)
 	topic := topicDispatcher.DispatchRowChangedEvent(row)
-	partition := partitionDispatcher.DispatchRowChangedEvent(row)
+	partition := partitionDispatcher.DispatchRowChangedEvent(row, partitionNum)
 
 	return topic, partition
 }
 
+// GetActiveTopics returns a list of the corresponding topics
+// for the tables that are actively synchronized.
+func (s *EventRouter) GetActiveTopics(activeTables []model.TableName) []string {
+	topics := make([]string, 0)
+	topicsMap := make(map[string]bool, len(activeTables))
+	for _, table := range activeTables {
+		topicDispatcher, _ := s.matchDispatcher(table.Schema, table.Table)
+		topicName := topicDispatcher.Substitute(table.Schema, table.Table)
```

Comment:
Maybe we can add a warning if the `topicName` generated by topic expression equals to `s.defaultTopic`.

Reply:
Should this be expected?

Reply:
> Should this be expected?

This is not expected. But for the current implementation, we cannot prevent users specifying a default topic which accidentally equals to `topicName`

Reply:
Or maybe we can add checks somewhere?

Reply:
Added. https://github.com/pingcap/tiflow/pull/4743/commits/20a9a413b85ddfde483e370b4ff457f0e223f9f0